File: src/main/java/uk/co/threebugs/mochiwhattotrade3/setup/SetupMapper.java
package uk.co.threebugs.mochiwhattotrade3.setup;

import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import uk.co.threebugs.mochiwhattotrade3.setupgroup.SetupGroup;

import java.nio.file.Path;

@Mapper(componentModel = "spring")
public interface SetupMapper {

    @Mapping(target = "setupGroup", source = "setupGroup")
    SetupDto toDto(Setup setup);

    @Mapping(target = "id", source = "setupDto.id")
    @Mapping(target = "symbol", source = "setupDto.symbol")
    Setup toEntity(SetupDto setupDto, SetupGroup setupGroup);


    default Path toPath(String path) {
        return Path.of(path);
    }

    default String toString(Path path) {
        return path.toString();
    }
}

File: src/main/kotlin/uk/co/threebugs/mochiwhattotrade3/setup/SetupMapper.kt
package uk.co.threebugs.mochiwhattotrade3.setup

import org.mapstruct.Mapper
import org.mapstruct.Mapping
import uk.co.threebugs.mochiwhattotrade3.setupgroup.SetupGroup
import java.nio.file.Path
import java.nio.file.Paths

@Mapper(componentModel = "spring")
abstract class SetupMapper {

    abstract fun toDto(setup: Setup): SetupDto

    @Mapping(target = "id", source = "setupDto.id")
    @Mapping(target = "symbol", source = "setupDto.symbol")
    abstract fun toEntity(setupDto: SetupDto, setupGroup: SetupGroup): Setup

    fun pathToString(path: Path?): String? {
        return path?.toString()
    }

    fun stringToPath(path: String?): Path? {
        return if (path == null) null else Paths.get(path)
    }
}
